### ISSUE TITLE
`hadrian/build.sh clean` continues building Hadrian, instead of cleaning stuff up

### ISSUE THREAD

**Issue by thomie** [snowleopard/hadrian#249]
`make clean` or `hadrian/build.sh clean` should preferably finish in a few seconds.

Scenario: a user has second-thoughts about starting their Hadrian-powered ghc build, or wants to still tweak some settings and start over.

The `clean` target should perhaps be implemented in a script.

**Comment by snowleopard** [comment]
A related issue: at the moment we are not cleaning the Hadrian executable (it cannot delete itself).

Once the move of all build results to `buildRootPath` is complete (#113) cleaning will be as simple as deleting the build directory.

**Comment by ndmitchell** [comment]
Is the goal of this to avoid compiling Hadrian itself? If so, I disagree - having the first clean take 1 min to compile Hadrian is no big deal, but having two sources of information or duplicating command line parsing is inadvisable.

**Comment by thomie** [comment]
> Is the goal of this to avoid compiling Hadrian itself?
yes

> having the first clean take 1 min to compile Hadrian is no big deal
it was a small annoyance to me

Priority lowest.

**Comment by snowleopard** [comment]
Just to record here that there may be more complexity to `clean` than I previously thought:
> Build system is IMO awful. It works fine, after a fresh clone ./boot && ./configure && ./make just works. But you can't:
* **Clean only stage2.**
* **Clean only libraries.**
* Build only stage1/2 from the top level.
* Build only the libraries.
* Uninstall GHC.

(From this reddit thread: https://www.reddit.com/r/haskell/comments/4isua9/ghc_development_outsidein/. Emphasis is mine.)

Cleaning only specific stages or only specific package(s) may be useful and easy implement in Hadrian, but I think it would be next to impossible to support this from within a script.

A script for cleaning everything (just by deleting `_build`) can be easily written by the user in their favourite script language. I don't think it is really worth to ship this trivial script together with the build system.

**Comment by izgzhen** [comment]
@snowleopard I guess this is a wontfix?

**Comment by snowleopard** [comment]
@izgzhen Let's keep this open for now. I've marked it low priority. Perhaps this will become more important when we have more Hadrian users.

**Comment by alpmestan** [comment]
A lot of this should be significantly easier with the new build-dir structure though, right?

**Comment by snowleopard** [comment]
Yes, clean up is pretty easy now.

A special script even feels superfluous, since it's going to be just something like `rm -rf _build`.

**Comment by alpmestan** [comment]
Right. Except for those files created by `boot` and `configure`, which still live in the source tree.

**Comment by snowleopard** [comment]
True, we need to keep track of files produced by `boot` and `configure`, which is a bit annoying.

**Comment by angerman** [comment]
If you really want to take care of boot and configure, I think the best option is to just invoke `git clean -xfd`. That of course requires a git checkout.

Maybe the question "What does one want to achieve with that clean step?" needs to be answered first.

**Comment by alpmestan** [comment]
Yeah, it should be answered first, because `git clean -xdf` would remove e.g new files that you have not yet registered in git.

**Comment by snowleopard** [comment]
I believe this issue is out of date, so closing.
